### PR TITLE
register node type

### PR DESCRIPTION
**Pull request by AsaAyers**
The first goal of this PR is to allow registering node types instead of implementing them all in `src/processor.js`. This was also a good time to rename `context` to `processingContext` to avoid confusion with React's context.

This seemed like a good opportunity to move `props` and `state` into `processingContext` as they're used by very few nodes.

I think it might make sense to move `...args` into the context, but that's going to require some more thought. I don't think passing `...args` through `group()` makes any sense, so that probably shouldn't work. `...args` moves unmodified through `lazy()`, but then are consumed by `loader()` and `selector()`.

### CODE REVIEW

**Comment by dcgleim** [comment]
I like this very much, especially moving the processing implementation in with the nodes.

`processorContext` is helpful, but since the thing I would confuse it with is still called `loaderContext`, I'm not sure they're the best.  I don't actually have better suggestions, and these aren't horrible.

`processorContext` beside is a bit of an awkward object, eh?  It looks like a pattern I see a lot here with more functional code where we "cheat" (except I don't know what the right alternative would be) and create a big bundle of state like a class might manage, but pass it around to functions that may or may not use all of it; which to my mind makes those functions effectively no different than methods.  `processorContext` contains a few things I am comfortable calling "context" or "parameters"--`state` and `props`--but then also a couple things that fall more in the stateful and output category: the method, `queue`, and the property `path`.

None of that is really in the scope of this PR, just taking the chance to get your thoughts.

File: src/node_types/depends.js [hunk 1]
```diff
@@ -24,32 +25,79 @@ function normalizeDependency(dep) {
...
+    },
+})
 
 depends.prop = prop
 depends.resource = resource
-
-// I don't know if `value` is useful outside of tests
 depends.value = value
 
 depends.TYPE = TYPE
```
Comment (dcgleim) [inline src/node_types/depends.js]: missed this one

File: src/node_types/group.js [hunk 2]
```diff
@@ -1,30 +1,70 @@
...
+        const resources = {}
+        const { path } = processingContext
+
+        const results = reduce(
+            node.children,
+            (x, child, key) => {
+                const childProcessingContext = {
+                    ...processingContext,
+                    resources,
+                    path: [...path, key],
```
Comment (dcgleim) [inline src/node_types/group.js]: What exactly is the role `processingContext.path` plays?  I can see it being built, but not sure what actually needs it.

**Comment by dcgleim** [comment]
btw, any reason for the switch to camelCase for branch names?

**Comment by AsaAyers** [comment]
`scope` is probably a better name for `processorContext`. Its purpose is to provide data to lower nodes. `scope` is never modified, new scopes with different data are passed down to child nodes. 

I'm sure this analogy isn't great, but I have always thought of this as similar to a compiler/interpreter. I considered making a syntax like http://graphql.org/, but it isn't worth the effort. JS functions "parse" the code `selector(something)` into a node in an Abstract Syntax Tree `{ TYPE: 'selector', selector: something }`. `processor` is the interpreter that executes the AST. In that context `scope` seems like the right name.

I think the `queue` is kind of cheating here because its the only call with a side effect. But the variable is owned by the top level processor call, so the system remains side effect free as a whole.

**Comment by AsaAyers** [comment]
No reason for the branch name. I don't think the branch name matters unless you have a convention of including the issue number or something.

**Comment by dcgleim** [comment]
It's a good analogy, I've been using the same one.

**Pull request merged**